File: _skels/python-flask-skel/app/seed.py
"""Seed default user accounts from environment variables at startup.

Creates a regular user and a superuser if the corresponding env vars
are set and the accounts do not already exist.  Uses the same
``hash_password`` helper as ``/api/auth/register`` so the stored hashes
are interchangeable.

Env vars
--------
USER_LOGIN, USER_EMAIL, USER_PASSWORD       — default regular user
SUPERUSER_LOGIN, SUPERUSER_EMAIL, SUPERUSER_PASSWORD — default superuser
"""

import os

from app import db
from app.auth import hash_password
from app.models import User


def seed_default_accounts() -> None:
    """Create default user and superuser accounts from env vars.

    Skips creation when the username already exists or the required env
    vars are not set.  Safe to call on every startup — existing rows are
    never modified.
    """

    _seed_account(
        username=os.environ.get("USER_LOGIN", "").strip(),
        email=os.environ.get("USER_EMAIL", "").strip(),
        password=os.environ.get("USER_PASSWORD", "").strip(),
    )

    _seed_account(
        username=os.environ.get("SUPERUSER_LOGIN", "").strip(),
        email=os.environ.get("SUPERUSER_EMAIL", "").strip(),
        password=os.environ.get("SUPERUSER_PASSWORD", "").strip(),
    )


def _seed_account(*, username: str, email: str, password: str) -> None:
    """Insert a single account if the username is provided and new."""

    if not username or not password:
        return

    existing = db.session.query(User).filter_by(username=username).first()
    if existing is not None:
        return

    user = User(
        username=username,
        email=email,
        password_hash=hash_password(password),
    )
    db.session.add(user)
    db.session.commit()

File: _skels/go-skel/internal/seed/seed.go
// Package seed creates default user accounts from environment variables
// at startup. Each account is only inserted when no row with the same
// username exists, so the function is safe to call on every boot.
package seed

import (
	"context"
	"errors"
	"log"
	"os"

	"gorm.io/gorm"

	"github.com/example/go-skel/internal/auth"
	"github.com/example/go-skel/internal/models"
)

// account describes a single default user to seed.
type account struct {
	loginEnv    string
	emailEnv    string
	passwordEnv string
	loginDef    string
	emailDef    string
	passwordDef string
}

// SeedDefaultAccounts reads USER_* and SUPERUSER_* env vars and
// inserts the corresponding rows into the users table when they do
// not already exist.
func SeedDefaultAccounts(ctx context.Context, db *gorm.DB) error {
	accounts := []account{
		{
			loginEnv:    "USER_LOGIN",
			emailEnv:    "USER_EMAIL",
			passwordEnv: "USER_PASSWORD",
			loginDef:    "user",
			emailDef:    "user@example.com",
			passwordDef: "secret",
		},
		{
			loginEnv:    "SUPERUSER_LOGIN",
			emailEnv:    "SUPERUSER_EMAIL",
			passwordEnv: "SUPERUSER_PASSWORD",
			loginDef:    "admin",
			emailDef:    "admin@example.com",
			passwordDef: "secret",
		},
	}

	tx := db.WithContext(ctx)
	for _, a := range accounts {
		login := envOr(a.loginEnv, a.loginDef)
		email := envOr(a.emailEnv, a.emailDef)
		password := envOr(a.passwordEnv, a.passwordDef)

		var existing models.User
		err := tx.Where("username = ?", login).First(&existing).Error
		if err == nil {
			log.Printf("seed: user %q already exists, skipping", login)
			continue
		}
		if !errors.Is(err, gorm.ErrRecordNotFound) {
			return err
		}

		hashed, err := auth.HashPassword(password)
		if err != nil {
			return err
		}
		user := models.User{
			Username:     login,
			Email:        email,
			PasswordHash: hashed,
		}
		if err := tx.Create(&user).Error; err != nil {
			return err
		}
		log.Printf("seed: created default user %q (%s)", login, email)
	}
	return nil
}

// envOr returns the value of the environment variable named key, or
// fallback when the variable is unset or empty.
func envOr(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}
